### PR TITLE
Add automatic type detection for Embedded.

### PR DESCRIPTION
**Pull request by Warxcell**
Great work on https://github.com/doctrine/orm/pull/8439 and https://github.com/doctrine/orm/pull/8589.
Following it - This adds same automatic type detection when using typed properties for Embedded. 

Needs help on PHPStan failure.

Edit: Figured it out.

### CODE REVIEW

File: lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php [hunk 1]
```diff
@@ -3605,9 +3605,16 @@ public function mapEmbedded(array $mapping)
     {
         $this->assertFieldNotMapped($mapping['fieldName']);
 
+        if (! isset($mapping['class']) && $this->isTypedProperty($mapping['fieldName'])) {
+            $type = $this->reflClass->getProperty($mapping['fieldName'])->getType();
+            if ($type) {
```
Comment (troytft) [inline lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php]:
```suggestion
            if ($type instanceof \ReflectionNamedType) {
```

to avoid  unsupported \ReflectionUnionType()

File: lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php [hunk 2]
```diff
@@ -3605,9 +3605,16 @@ public function mapEmbedded(array $mapping)
...
+            $type = $this->reflClass->getProperty($mapping['fieldName'])->getType();
+            if ($type instanceof ReflectionNamedType) {
+                $mapping['class'] = $type->getName();
+            }
+        }
+
         $this->embeddedClasses[$mapping['fieldName']] = [
             'class' => $this->fullyQualifiedClassName($mapping['class']),
-            'columnPrefix' => $mapping['columnPrefix'],
+            'columnPrefix' => $mapping['columnPrefix'] ?? null,
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:3605]: :thinking: what does this have to do with the rest of your PR?

File: lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php [hunk 3]
```diff
@@ -322,9 +322,11 @@ public function loadMetadataForClass($className, ClassMetadata $metadata)
                     ? $this->evaluateBoolean($embeddedMapping['use-column-prefix'])
                     : true;
 
+                $class = isset($embeddedMapping['class']) ? (string) $embeddedMapping['class'] : null;
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php]: Let's inline this on line 327 and get rid of this one-time variable?

File: lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php [hunk 4]
```diff
@@ -3605,9 +3605,16 @@ public function mapEmbedded(array $mapping)
     {
         $this->assertFieldNotMapped($mapping['fieldName']);
 
+        if (! isset($mapping['class']) && $this->isTypedProperty($mapping['fieldName'])) {
+            $type = $this->reflClass->getProperty($mapping['fieldName'])->getType();
+            if ($type instanceof ReflectionNamedType) {
+                $mapping['class'] = $type->getName();
+            }
+        }
+
```
Comment (beberlei) [inline lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:3602]: You need to check here again that class isset or otherwise throw an exception. If not, then line 3616 will throw a notice accessing mapping class that does not exist.

File: lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php (hunk 2 above)
Comment (beberlei) [reply lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:3605]: I assume phpstan failure they mentioned in the PR description.

File: lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php (hunk 4 above)
Comment (beberlei) [reply lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:3602]: Actually the validation is done "late" in `ClassMetadataFactory::doLoadMetadata` as per https://github.com/doctrine/orm/pull/1133 - this should probably be changed, but not now.

**Review by beberlei — APPROVED** [review]

**Comment by Warxcell** [comment]
@greg0ire since columnPrefix can be null, it makes no sense to make it required and pass null. So I just allow to omit it.
